File: extensions/youtube/src/main/java/app/morphe/extension/youtube/patches/OverrideYouTubeMusicButtonsPatch.java
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 Section 7 terms that apply to this code.
 */

package app.morphe.extension.youtube.patches;

import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import app.morphe.extension.shared.Utils;
import app.morphe.extension.youtube.settings.Settings;

@SuppressWarnings("unused")
public class OverrideYouTubeMusicButtonsPatch {

    private static final String YOUTUBE_MUSIC_PACKAGE_NAME = "com.google.android.apps.youtube.music";
    private static final String MORPHE_MUSIC_PACKAGE_NAME = "app.morphe.android.apps.youtube.music";
    private static final String HIJACK_FLAG = "morphe_hijacked";

    private static String getTargetPackage() {
        String customTarget = Settings.CUSTOM_MUSIC_PACKAGE_NAME.get().trim();
        if (customTarget.isEmpty()) {
            return MORPHE_MUSIC_PACKAGE_NAME;
        }
        return customTarget;
    }

    private static Intent hijackIntent(Intent intent) {
        String target = getTargetPackage();
        PackageManager pm = Utils.getContext().getPackageManager();
        Intent launchIntent = pm.getLaunchIntentForPackage(target);

        if (launchIntent == null) {
            intent.setData(Uri.parse("https://music.youtube.com/"));
            intent.setPackage(null);
            intent.setComponent(null);
            return intent;
        }

        Uri data = intent.getData();
        String uriString = data != null ? data.toString() : "";
        boolean isPlayStoreLink = uriString.contains("play.google.com") || uriString.startsWith("market://");

        intent.setAction(launchIntent.getAction());
        intent.setComponent(launchIntent.getComponent());
        intent.setPackage(target);
        intent.putExtra(HIJACK_FLAG, true);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        if (isPlayStoreLink || data == null) {
            intent.setData(null);
        } else {
            intent.setData(data);
        }

        return intent;
    }

    public static Intent overrideSetPackage(Intent intent, String packageName) {
        if (intent == null) return null;

        if (!Settings.OVERRIDE_YOUTUBE_MUSIC_BUTTONS.get()) {
            return intent.setPackage(packageName);
        }

        if (intent.getBooleanExtra(HIJACK_FLAG, false)) {
            return intent;
        }

        if (YOUTUBE_MUSIC_PACKAGE_NAME.equals(packageName)) {
            return hijackIntent(intent);
        }

        return intent.setPackage(packageName);
    }

    public static Intent overrideSetData(Intent intent, Uri uri) {
        if (intent == null) return null;
        if (uri == null) return intent.setData(null);

        if (!Settings.OVERRIDE_YOUTUBE_MUSIC_BUTTONS.get()) {
            return intent.setData(uri);
        }

        if (intent.getBooleanExtra(HIJACK_FLAG, false)) {
            return intent;
        }

        String uriString = uri.toString();
        if (uriString.contains(YOUTUBE_MUSIC_PACKAGE_NAME) || uriString.contains("music.youtube.com")) {
            intent.setData(uri);
            return hijackIntent(intent);
        }

        return intent.setData(uri);
    }

    public static Intent overrideSetComponent(Intent intent, ComponentName component) {
        if (intent == null) return null;

        if (!Settings.OVERRIDE_YOUTUBE_MUSIC_BUTTONS.get()) {
            return intent.setComponent(component);
        }

        if (intent.getBooleanExtra(HIJACK_FLAG, false)) {
            return intent;
        }

        if (component != null && YOUTUBE_MUSIC_PACKAGE_NAME.equals(component.getPackageName())) {
            intent.setComponent(component);
            return hijackIntent(intent);
        }

        return intent.setComponent(component);
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/layout/music/OverrideYouTubeMusicButtonsPatch.kt
/*
 * Copyright 2026 Morphe.
 * https://github.com/MorpheApp/morphe-patches
 *
 * See the included NOTICE file for GPLv3 §7(b) and §7(c) terms that apply to this code.
 */

package app.morphe.patches.youtube.layout.music

import app.morphe.patcher.extensions.InstructionExtensions.replaceInstruction
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patcher.patch.resourcePatch
import app.morphe.patches.shared.misc.settings.preference.PreferenceCategory
import app.morphe.patches.shared.misc.settings.preference.PreferenceScreenPreference.Sorting
import app.morphe.patches.shared.misc.settings.preference.SwitchPreference
import app.morphe.patches.shared.misc.settings.preference.TextPreference
import app.morphe.patches.youtube.misc.settings.PreferenceScreen
import app.morphe.patches.youtube.misc.settings.settingsPatch
import app.morphe.patches.youtube.shared.Constants.COMPATIBILITY_YOUTUBE
import app.morphe.util.findMutableMethodOf
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.instruction.FiveRegisterInstruction
import com.android.tools.smali.dexlib2.iface.instruction.ReferenceInstruction
import com.android.tools.smali.dexlib2.iface.instruction.RegisterRangeInstruction
import com.android.tools.smali.dexlib2.iface.reference.MethodReference

private const val EXTENSION_CLASS = "Lapp/morphe/extension/youtube/patches/OverrideYouTubeMusicButtonsPatch;"

private fun overrideYouTubeMusicManifestPatch() = resourcePatch{
    compatibleWith(COMPATIBILITY_YOUTUBE)

    execute {
        val manifestFile = get("AndroidManifest.xml")
        var manifestContent = manifestFile.readText()
        val permissionTag = "<uses-permission android:name=\"android.permission.QUERY_ALL_PACKAGES\"/>"

        if (!manifestContent.contains(permissionTag)) {
            manifestContent = manifestContent.replace(
                "<application",
                "$permissionTag\n    <application"
            )
            manifestFile.writeText(manifestContent)
        }
    }
}

@Suppress("unused")
val overrideYouTubeMusicButtonsPatch = bytecodePatch(
    name = "Override YouTube Music buttons",
    description = "Overrides YouTube Music buttons to open Morphe Music or any compatible third-party client.",
) {
    dependsOn(settingsPatch)
    dependsOn(overrideYouTubeMusicManifestPatch())
    compatibleWith(COMPATIBILITY_YOUTUBE)

    execute {
        PreferenceScreen.GENERAL.addPreferences(
            PreferenceCategory(
                titleKey = null,
                sorting = Sorting.UNSORTED,
                tag = "app.morphe.extension.shared.settings.preference.NoTitlePreferenceCategory",
                preferences = setOf(
                    SwitchPreference(key = "morphe_override_youtube_music_buttons"),
                    TextPreference(key = "morphe_music_package_name")
                )
            )
        )

        classDefForEach { classDef ->
            if (classDef.type == EXTENSION_CLASS) return@classDefForEach
            var needsPatch = false
            classDef.methods.forEach { method ->
                if (method.implementation?.instructions?.any {
                        it.opcode == Opcode.INVOKE_VIRTUAL &&
                                (it as? ReferenceInstruction)?.reference?.let { ref ->
                                    val mRef = ref as? MethodReference
                                    mRef?.definingClass == "Landroid/content/Intent;" &&
                                            (mRef.name == "setPackage" || mRef.name == "setData")
                                } == true
                    } == true) {
                    needsPatch = true
                }
            }

            if (needsPatch) {
                val mutableClass = mutableClassDefBy(classDef.type)
                classDef.methods.forEach { method ->
                    val instructions = method.implementation?.instructions?.toList() ?: return@forEach
                    val targetIndices = instructions.mapIndexedNotNull { index, instruction ->
                        if (instruction.opcode == Opcode.INVOKE_VIRTUAL) {
                            val ref = (instruction as? ReferenceInstruction)?.reference as? MethodReference
                            if (ref?.definingClass == "Landroid/content/Intent;") {
                                if (ref.name == "setPackage" && ref.parameterTypes == listOf("Ljava/lang/String;")) {
                                    index to "overrideSetPackage(Landroid/content/Intent;Ljava/lang/String;)Landroid/content/Intent;"
                                } else if (ref.name == "setData" && ref.parameterTypes == listOf("Landroid/net/Uri;")) {
                                    index to "overrideSetData(Landroid/content/Intent;Landroid/net/Uri;)Landroid/content/Intent;"
                                } else null
                            } else null
                        } else null
                    }

                    if (targetIndices.isNotEmpty()) {
                        val mutableMethod = mutableClass.findMutableMethodOf(method)
                        targetIndices.reversed().forEach { (index, methodDescriptor) ->
                            val instruction = instructions[index]
                            val invokeString = if (instruction is RegisterRangeInstruction) {
                                "invoke-static/range {v${instruction.startRegister} .. v${instruction.startRegister + instruction.registerCount - 1}}"
                            } else {
                                val i = instruction as FiveRegisterInstruction
                                "invoke-static {v${i.registerC}, v${i.registerD}}"
                            }

                            mutableMethod.replaceInstruction(
                                index,
                                "$invokeString, $EXTENSION_CLASS->$methodDescriptor"
                            )
                        }
                    }
                }
            }
        }
    }
}